### ISSUE TITLE
Serializing Int64 in 2022.3.4f1 creates a different value due to inlining

### ISSUE THREAD

**Issue by oddgames-david** [TeamSirenix/odin-serializer#59]
We've had to do this to fix it.

//[MethodImpl((MethodImplOptions)0x100)]  // Set aggressive inlining flag, for the runtimes that understand that
[MethodImpl(MethodImplOptions.NoInlining)]
private bool UNSAFE_Read_8_Int64(out long value)

Here's the same old serialized data being loaded on unity 2021.3.24 and unity 2022.3.4
[load from 31460.zip](https://github.com/TeamSirenix/odin-serializer/files/11955620/load.from.31460.zip)
[load from 31647.zip](https://github.com/TeamSirenix/odin-serializer/files/11955619/load.from.31647.zip)

**Comment by Syloo** [comment]
On my Windows PC everything works fine, but when using Odin in an Android app then I have serious problems with `long` numbers.
As an example, I serialize `5249950202304764354` with Odin, then deserialize it back and get `540106660290`.
I switched from Unity 2021.03.24f1, where everything seemed to work fine, to Unity 2022.03.5f1. In my Unity project settings I am using .Net Framework with IL2CPP and a target Android API level 33 and minimum Android API level 22.
This caused a feature of my app to not work properly and I have heard that it has the same problem on iPhones as well.
@oddgames-david Sadly your fix did not work for me.

I hope this gets fixed pretty soon, because more and more people are upgrading to Unity 2021+ to be able to build for higher Android API levels which are getting enforced by the Google Play Store.

**Comment by Syloo** [comment]
I have now found a solution that works for me. (Maybe I should add that I build to target architectures ARMv7 and ARM64)

Here are the minimal changes I did to the OdinSerializer code to make it work:
1. I changed the `UNSAFE_WriteToBuffer_8_Int64(long value)` in BinaryDataWriter.cs to use MethodImplOptions.NoInlining instead, just like @oddgames-david (note that I modified the unsafe write, not the unsafe read function).
2. Also in `UNSAFE_WriteToBuffer_8_Int64(long value)` I changed
```
*toPtr++ = *fromPtr++;
*toPtr = *fromPtr;
```
to
```
*toPtr = *fromPtr++;
toPtr++;
*toPtr = *fromPtr;
```
or
```
*toPtr = *fromPtr++;
*(toPtr + 1) = *fromPtr;
```
or even
```
*(toPtr++) = *fromPtr++;
*toPtr = *fromPtr;
```

I am not really sure, why my code works and the OdinSerializer code doesn't, because they are logically the same.
I noticed that for the `long` numbers I tested, the first 32 of 64 bits were always correctly serialized and deserialized (the 32 least significant bits because of little-endian). For the second 32 bits (most significant bits) the first 3 bytes were zero and the 4th was seemingly random. This caused the serialized and then deserialized number to be magnitudes smaller than the original.

My first guess was that the IL2CPP conversion of Unity 2022.x messes something up and `*toPtr++ = *fromPtr++;` does not increase the address of `int*` pointer `toPtr` by 4 as it should but instead by 1, because `toPtr` is constructed from a `byte*` pointer.
```
// Types: long value, byte* basePtr, int this.bufferIndex
int* fromPtr = (int*)&value;
int* toPtr = (int*)(basePtr + this.bufferIndex);
```
Then the second write to `toPtr` would be unaligned which seems to lead to undefined behaviour (because unaligned access apparently is not supported on ARMv7). However, I computed the address offset between the original and incremented `toPtr` but it was always 4 both for the working and the broken code.

Someone who knows how to inspect the intermediate language/C++ code generated from Unitys IL2PP conversion will probably find differences between the existing, non-working code and my slightly modified, working code.
I am pretty sure IL2CPP is causing this error, because **this problem does not occur when using Mono** with the unmodified OdinSerializer.

**Comment by oddgames-david** [comment]
Apologies, I actually did both reader and writer in my working version.

**Comment by TorVestergaard** [comment]
Huh. I came back from vacation last week and then spent the entire week debugging and solving this issue, and only just now worked through enough of my email backlog that I found this. So I entirely missed this breakdown where you discovered the cause before I ever did!

Well, a retroactive thank you both very much, here! - This was a pain to debug myself, and I imagine it must have been a pain for you as well. As of patch 3.1.14.3, this is fixed in Odin Inspector, and I will be pushing the necessary changes to this repo throughout next week. They amount to basically exactly what you discovered, namely that all methods that write larger than 8 byte values are now marked as non-inlinable.

**Comment by Syloo** [comment]
I am happy that this is getting fixed. Odin Serializer is a great tool when it works :)
It seems that for the both of you disabling inlining was enough to resolve the issue. However, I had to do one of the above modifications on top of that to make it work, so I did 2 additional tests:

1. I tested the serialization of a long when building with a linux system instead of windows. I used the same android phone as before to execute the apk => Changing the write function to no inline was enough to fix the problem
2. I removed OdinSerializer from my project on windows (and temporarly fixed the errors this caused), then downloaded the buggy (current) version, imported the unitypackage into the project and tested it again => Changing the write function to no inline was enough to fix the problem

Also I tracked all the files in the `Assets` folder before and after re-installing Odin Serializer and nothing changed. I guess I had some strange cached data or other weird internal state that caused the inlining fix to not be enough to resolve the problem.
I hope that the update process goes well for other people and they don't have some additional strange behaviour as I had.

Maybe someone should inform the IL2CPP devs about the inlining problem, because it doesn't seem like this behaviour is intended. Is it Unity themselves who are responsible for this?

**Comment by TorVestergaard** [comment]
It would appear to be changes Unity made to IL2CPP that are responsible for this, yes. I need to sit down and take some time to send them a decent repro setup for this, though it seems we will have to leave the no inlining flags on to ensure maximum compatibility across affected versions of Unity. If somebody else wants to do that, too, then please - the more reports, the more weight the issue has.

As for whether the fix works, quite a few affected customers building on a wide variety of platforms and hardware have confirmed that it works, by now, with none saying it doesn't - so it seems like it's solid enough. We'll see as it comes into wider circulation if that trend holds true, but I'm hopeful.

**Comment by BioeJD** [comment]
I haven't run into this issue yet, but I am currently working on updating our app to Unity 2022 and there was another serialization issue I resolved by updating to the most current version available here: [https://odininspector.com/download](https://odininspector.com/download)

Is the issue in this thread fixed in the OdinSerializer version downloaded there, or would I need to pull from the repo?

**Comment by nodrock** [comment]
As I can see this serialization issue was fixed half a year ago but is not in odin serializer project only in inspector.
Could you add fix to that issue (https://bitbucket.org/sirenix/odin-inspector/issues/913/deserializevalue-reached-end-of-array) to open source odin serializer as well?

**Comment by TorVestergaard** [comment]
The original Odin Inspector fix ended up not working properly. It fixed this issue in some cases on some platforms. However, recently I managed to figure out a solution during a dedicated debugging session with a reproducing customer. The final fixes are in the commit 25bef7cdbb21858cf59db8e22c7a720ca90d2ebc which will also land in an Odin Inspector patch very soon.
